### PR TITLE
Allow blog to use APIs from different environments

### LINKED ISSUE #308: Use the proper API for each environment
This is a follow-up to https://github.com/mozilla/addons-blog/issues/298.

We should be using the dev API for the dev instance of the blog, and the stage API for the stage instance of the blog. One complication is that we'll want the add-ons for an article to work in each environment, but the `id`s are not currently in sync.

Ideally we should be using `guid` for this, as opposed to `addonId`, because the latter is the primary key to a table and therefore more difficult to change.

We need to check the code to see what is currently being used, and if it is currently `addonId` look into how difficult it would be to change it to use `guid` instead.

### PR DESCRIPTION
**Pull request by bobsilverberg**
~~Depends on https://github.com/mozilla/addons-blog/pull/343~~
Fixes #308

### CODE REVIEW

**Comment by bobsilverberg** [comment]
I still need to add tests for the changes for the dynamic script. I'm not sure what to do about the change to the `body` tag. Should a test be added to https://github.com/mozilla/addons-blog/blob/main/tests/php/WordPressThemeTest.php for that? Also, how does one run the php tests locally. I've never actually used php before, believe it or not!

File: src/filters.js
Comment (willdurand) [inline src/filters.js:87]: Using negate `\s` (so anything but (white)spaces) does not sound like a great idea because it isn't very readable. I'd suggest either `.+?` (non-greedy) or alphanumeric + `-` but not sure that is necessary.

**Review by willdurand — COMMENTED** [review]
Is this going to work right away for all envs? If so, r- because I'd like -stage and -prod to be identical until there is a better (as in safer) way to deploy new articles to the -prod.

**Comment by bobsilverberg** [comment]
> Is this going to work right away for all envs? If so, r- because I'd like -stage and -prod to be identical until there is a better (as in safer) way to deploy new articles to the -prod.

I hadn't thought about the fact that the `AMO_BASE_URL` environment variable is already being set differently for dev, stage, and prod (I assume), so in that case, yes, this would affect all environments. I agree we should exclude stage from this so I see a few options:
1. Introduce a separate environment variable for `AMO_API_URL`.
2. Introduce a new environment variable to control whether to override the default URL for the API or not. This would introduce a boolean env var, as opposed to basically duplicating the existing `AMO_BASE_URL`.
3. Special case the code as it stands in this patch to exclude stage from using the environment variable for the API.

I think they all have their downsides, but 3 might be best because it doesn't introduce another environment variable.

@willdurand WDYT?

**Comment by willdurand** [comment]
How wouldyou implement (3)? By looking at the value of `AMO_BASE_URL`?

The env vars are defined in the circle config file, I am fine with adding new env vars if that makes things crystal clear.

I am not against (3), though.

**Comment by bobsilverberg** [comment]
> How wouldyou implement (3)? By looking at the value of `AMO_BASE_URL`?

Yes, exactly. Then we can tell if we're currently on stage or not.

> 
> The env vars are defined in the circle config file, I am fine with adding new env vars if that makes things crystal clear.
> 
> I am not against (3), though.

3 seems like the simplest approach, and we'll likely keep things that way (with stage being excluded) for the forseeable future, so I don't think it will require much maintenance. I'll take that approach. 

I'm still trying to figure out how to test the changes to `src/layouts/base.njk` and the associated change to `src/assets/js/dynamic-addon-cards.js`. For the latter I was trying to inject my own value for `data-amo-base-url` into the `body` and have the test pick it up, but that doesn't seem to be working. I'll keep trying some variations of that, but do you have any suggestions for the change to `src/layouts/base.njk`?

**Comment by willdurand** [comment]
> I think MacOS has PHP installed by default, from there:

Looks like MacOS might not bundle `php` anymore these days. I had to install it on my new machine, so here is the commands I typed for (0) install PHP and (1) install composer:

```
brew install php
php --version
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php -r "if (hash_file('sha384', 'composer-setup.php') === '906a84df04cea2aa72f40b5f787e49f22d4c2f19492ac310e8cba5b96ac8b64115ac402c8cd292b8a03482574915d1a8') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;"
php composer-setup.php
php -r "unlink('composer-setup.php');"
```

**Comment by bobsilverberg** [comment]
@willdurand I added a couple of tests to `tests/php/WordPressThemeTest.php` for the change to `base.njk`, but they are not ideal. They do check that a `data-amo-base-url` attribute is set on the `body` tag, but they only work with the default value, and don't verify that this value can be picked up from an environment variable.

From what I can tell, the environment variable is picked up during the running of `yarn build:wptheme`, not during the running of the actual test, so setting an environment variable for the test run doesn't have any effect. I can force the environment variable to be something when running `yarn build:wptheme`, and then check for that same value in the test, but that seems fragile to me. I guess it would all need to be set up in the `test-wptheme` job in the `.circleci/config.yml` file.  I suppose that if I can define an environment variable for that job, and then use it in both the running of `yarn build:wptheme` as well as `./vendor/bin/phpunit`, then it wouldn't be so fragile. I'll give that a try. Do you agree that is what we should do?

**Comment by bobsilverberg** [comment]
Hmm, I see I also accidentally committed and pushed some changes I made while trying to write another test, which are still a work in progress. Please ignore those for now.

**Comment by willdurand** [comment]
> Hmm, I see I also accidentally committed and pushed some changes I made while trying to write another test, which are still a work in progress. Please ignore those for now.

----- ok, np 😁

**Comment by bobsilverberg** [comment]
> @willdurand I added a couple of tests to `tests/php/WordPressThemeTest.php` for the change to `base.njk`, but they are not ideal. They do check that a `data-amo-base-url` attribute is set on the `body` tag, but they only work with the default value, and don't verify that this value can be picked up from an environment variable.
> 
> From what I can tell, the environment variable is picked up during the running of `yarn build:wptheme`, not during the running of the actual test, so setting an environment variable for the test run doesn't have any effect. I can force the environment variable to be something when running `yarn build:wptheme`, and then check for that same value in the test, but that seems fragile to me. I guess it would all need to be set up in the `test-wptheme` job in the `.circleci/config.yml` file. I suppose that if I can define an environment variable for that job, and then use it in both the running of `yarn build:wptheme` as well as `./vendor/bin/phpunit`, then it wouldn't be so fragile. I'll give that a try. Do you agree that is what we should do?

Thinking about this some more, I'm not sure that setting a test environment variable in the `.circleci/config.yml` is a very good approach. Something just feels wrong about that. Do you have any other suggestions for how we might create a test for `base.njk` that allows us to verify that it picks up an environment variable if one exists?

**Comment by willdurand** [comment]
> Thinking about this some more, I'm not sure that setting a test environment variable in the `.circleci/config.yml` is a very good approach. Something just feels wrong about that. Do you have any other suggestions for how we might create a test for `base.njk` that allows us to verify that it picks up an environment variable if one exists?

The alternative would be to add some tests on the production build: https://github.com/mozilla/addons-blog/blob/main/tests/integration-production.test.js but we'd need to have the right value in an env var as well.

**Comment by bobsilverberg** [comment]
This is (finally) ready for a review, @willdurand. Sorry for the delay.

File: .circleci/config.yml
Comment (willdurand) [inline .circleci/config.yml:69]: Maybe add a comment to explain why it is needed here.

File: .gitignore
Comment (willdurand) [inline .gitignore:3]: Not sure we need that but OK.

File: src/assets/js/dynamic-addon-cards.js
Comment (willdurand) [inline src/assets/js/dynamic-addon-cards.js:80]: Is this really needed? I am fine with it but, in theory, the Eleventy build step should already perform this check?

File: src/assets/js/dynamic-addon-cards.js
Comment (willdurand) [inline src/assets/js/dynamic-addon-cards.js:75]: I suppose it is fine here but this could be done only once at the bottom of this file, and we would pass `baseURL` or `baseURLFromBody` as an optional argument (like `parsedUserAgent`).

File: tests/php/WordPressThemeTest.php
Comment (willdurand) [inline tests/php/WordPressThemeTest.php:28]: Why is this needed?

File: tests/php/WordPressThemeTest.php
Comment (willdurand) [inline tests/php/WordPressThemeTest.php:44]: I think you can do:

```suggestion
            "<body data-amo-base-url=\"{self::$AMO_BASE_URL}\">",
```

File: .circleci/config.yml
Comment (bobsilverberg) [reply .circleci/config.yml:69]: Rather than trying to explain it in a comment, I included a comment that points back to the discussion in the PR.

File: .gitignore
Comment (bobsilverberg) [reply .gitignore:3]: What are the other options? Include it in the repo, or delete it from my machine before committing? Or, I suppose, only add it to my own `.gitignore`? This seemed like the simplest and least effort of those options, and could help others who end up running the tests locally.

File: src/assets/js/dynamic-addon-cards.js
Comment (bobsilverberg) [reply src/assets/js/dynamic-addon-cards.js:80]: Are you sure? The way I was understanding it is that the Eleventy build step creates the initial versions of the cards, which does include a version of this logic, but the code in `dynamic-addon-cards.js` runs in the browser to update the cards, and it picks up the `baseURL` from the `body`, which would still be the value for the specific environment. Without a check here, wouldn't we be using the stage API when we are on stage, which is what we didn't want?

File: src/assets/js/dynamic-addon-cards.js
Comment (bobsilverberg) [reply src/assets/js/dynamic-addon-cards.js:75]: If I do that then the tests I wrote no longer work, because they are designed to test reading the `baseURL` from the body, but with this design we actually pass the `baseURL` into the call to `updateAddonCard`, so we would need a different set of tests. I actually moved this code from the top of this file into `updateAddonCard` in order to be able to get these tests to work. Does that seem ok?

File: tests/php/WordPressThemeTest.php
Comment (bobsilverberg) [reply tests/php/WordPressThemeTest.php:28]: I added this so we can verify that the test did in fact run with the expected environment variable. Without it, there is no way to know if our intent to run the test with an `AMO_BASE_URL` for dev actually happened. I guess we don't need it on CI, after I have verified it for myself locally, but it seemed worth keeping in place.

Comment (bobsilverberg) [reply tests/php/WordPressThemeTest.php:28]: Actually, I can't verify the behaviour on CI locally, so this is still needed if we want to be sure it's working (and there wasn't an error in the CI configuration).

File: tests/php/WordPressThemeTest.php
Comment (bobsilverberg) [reply tests/php/WordPressThemeTest.php:44]: That doesn't seem to work, but `"<body data-amo-base-url=\"$AMO_BASE_URL\">"` does, but only after I copy `self::$AMO_BASE_URL` into `$AMO_BASE_URL`. I couldn't seem to figure out how to use `self::$AMO_BASE_URL` in a template string. I'll update the patch with the change I just described.

File: .gitignore
Comment (willdurand) [reply .gitignore:3]: I personally have `composer` installed globally (in `/usr/local/bin/`) and for other cases, I use a global `.gitignore` file. It's fine, though.

File: src/assets/js/dynamic-addon-cards.js
Comment (willdurand) [reply src/assets/js/dynamic-addon-cards.js:80]: Right, so we should inject the correct value in the data attribute. Instead of using `markup.baseURL`, we should use `markup.baseApiUrl` probably.

Comment (bobsilverberg) [reply src/assets/js/dynamic-addon-cards.js]: I changed a bunch of code to do this, which allowed the logic for dev to reside in one place. This is ready for another look.

File: tests/php/WordPressThemeTest.php
Comment (willdurand) [reply tests/php/WordPressThemeTest.php:28]: mmm how about we use the phpunit config to inject the env var then? We can add the following XML in the [`phpunit.xml.dist`](https://github.com/mozilla/addons-blog/blob/main/phpunit.xml.dist) file:

```xml
<phpunit colors="true" bootstrap="tests/php/bootstrap.php">
  <testsuites>
    <testsuite name="AMO Blog Test Suite">
      <directory>tests/php/</directory>
    </testsuite>
  </testsuites>
  <php>
    <env name="AMO_BASE_URL" value="https://addons-dev.allizom.org" />
  </php>
</phpunit>
```

Comment (willdurand) [reply tests/php/WordPressThemeTest.php:28]: We can also add an assertion on the length of `$_ENV['AMO_BASE_URL']` to prevent empty values.

Comment (bobsilverberg) [reply tests/php/WordPressThemeTest.php:28]: Making that change causes the tests to fail locally, because the env var isn't set for `build:wptheme`. I could add it to the command for `build:wptheme`, but I'm not sure that's what we want. We could also include instructions for running `build:wptheme` with the env var set when testing locally, but that seems a bit complicated to me. We could add another `yarn` script that does what `build:wptheme` does, and includes the env var, and update the instructions to run *that* locally, but again, this seems overly complicated.

I think the set up was better before making that change, as it worked on both CI and locally, without having to force the env var locally.

WDYT?

Comment (willdurand) [reply tests/php/WordPressThemeTest.php:28]: Gaah, I forgot about that, _again_.

Because I don't quite like this code (which isn't too PHP-y), I'd suggest to print the variable in the shell script here: https://github.com/mozilla/addons-blog/blob/9e6e4c37c8f949c7bf0879785df9801b6ddbac89/.circleci/config.yml#L63-L66

Something like:

```sh
echo "AMO_BASE_URL=$AMO_BASE_URL"
```

**Review by willdurand — APPROVED** [review]
r+wc

**Comment by bobsilverberg** [comment]
@willdurand Thanks for all of your help with this. I had marked the issue with "do not merge", because I didn't want to risk breaking the blog if we need to deploy for a new post. I believe the last post of the year went out today, but I still think it would be prudent to wait until the new year to merge this.

**Comment by bobsilverberg** [comment]
@willdurand I think this is safe to merge now, so I will do so. We should be able to test it against the dev API once it's been deployed to dev.

**Pull request merged**
